### PR TITLE
Query TaskReschedule only if task is UP_FOR_RESCHEDULE

### PR DESCRIPTION
**Pull request by turbaszek**
---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Review by potiuk — COMMENTED** [review]
Nice optimisation :)

**Review by potiuk — APPROVED** [review]

**Comment by mik-laj** [comment]
@seelmann Can you look at it?

**Comment by seelmann** [comment]
Makes perfect sense, +1.

Another optimization could be to not fetch all `TaskReschedule` objects but only the first one. Similar in `ready_to_reschedule.py` only the last one is needed. I wanted to change that long time ago but didn't fine the time and then forgot...

**Comment by turbaszek** [comment]
> Another optimization could be to not fetch all `TaskReschedule` objects but only the first one. Similar in `ready_to_reschedule.py` only the last one is needed. 

Sounds reasonable, thanks!

**Comment by turbaszek** [comment]
@seelmann would like to take another look?

File: airflow/models/taskreschedule.py [hunk 1]
```diff
@@ -63,24 +63,41 @@ def __init__(self, task, execution_date, try_number, start_date, end_date,
...
             .query(TR)
             .filter(TR.dag_id == task_instance.dag_id,
                     TR.task_id == task_instance.task_id,
                     TR.execution_date == task_instance.execution_date,
                     TR.try_number == task_instance.try_number)
-            .order_by(asc(TR.id))
-            .all()
         )
+        if descending:
+            return qry.order_by(desc(TaskReschedule.id))
```
Comment (seelmann) [inline airflow/models/taskreschedule.py]:
```suggestion
            return qry.order_by(desc(TR.id))
```

File: airflow/models/taskreschedule.py [hunk 2]
```diff
@@ -63,24 +63,41 @@ def __init__(self, task, execution_date, try_number, start_date, end_date,
...
+        else:
+            return qry.order_by(asc(TaskReschedule.id))
```
Comment (seelmann) [inline airflow/models/taskreschedule.py]:
```suggestion
            return qry.order_by(asc(TR.id))
```

**Review by seelmann — COMMENTED** [review]
LGTM

**Comment by turbaszek** [comment]
CI error: `AssertionError: The expected number of db queries is 131. The current number is 135.`

File: airflow/ti_deps/deps/ready_to_reschedule.py [hunk 3]
```diff
@@ -52,14 +52,14 @@ def _get_dep_statuses(self, ti, session, dep_context):
                 reason="The task instance is not in State_UP_FOR_RESCHEDULE or NONE state.")
             return
 
-        task_reschedules = TaskReschedule.find_for_task_instance(task_instance=ti)
-        if not task_reschedules:
+        task_reschedule = TaskReschedule.query_for_task_instance(task_instance=ti, descending=True).first()
```
Comment (mik-laj) [inline airflow/ti_deps/deps/ready_to_reschedule.py]:
```suggestion
        task_reschedule = (
            TaskReschedule.query_for_task_instance(task_instance=ti, descending=True)
            .with_entities(TR.reschedule_date)
            .first()
        )
```

Comment (ashb) [reply airflow/ti_deps/deps/ready_to_reschedule.py]: Oh does `with_entites` say "just select this one column? Nice!

Comment (mik-laj) [reply airflow/ti_deps/deps/ready_to_reschedule.py]: This allows you to change the content of the parameter passed to the query method on an existing object. We already use it in the project. https://github.com/apache/airflow/pull/8729/files

Comment (mik-laj) [reply airflow/ti_deps/deps/ready_to_reschedule.py]: Here is doc: https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.with_entities

Comment (ashb) [inline airflow/ti_deps/deps/ready_to_reschedule.py]: Please pass on the same session object.
```suggestion
        task_reschedule = (
            TaskReschedule.query_for_task_instance(task_instance=ti, descending=True, session=session)
            .with_entities(TR.reschedule_date)
            .first()
        )
```

**Comment by turbaszek** [comment]
@ashb @mik-laj your suggestions are added

File: airflow/ti_deps/deps/ready_to_reschedule.py (hunk 3 above)
Comment (ashb) [reply airflow/ti_deps/deps/ready_to_reschedule.py]: You missed Kamil's suggestion of `.with_entites` here.

Comment (turbaszek) [reply airflow/ti_deps/deps/ready_to_reschedule.py]: I think it's there :)

**Review by mik-laj — APPROVED** [review]

**Review by ashb — APPROVED** [review]

**Comment by mik-laj** [comment]
@turbaszek Should we merge it?

**Pull request merged**